src/BasicInputs/DatetimeInput.tsx

```diff
@@ -1,27 +1,149 @@
+'use client';
+
+import { add, format } from 'date-fns';
+import { Calendar as CalendarIcon } from 'lucide-react';
 import {
-  type ComponentPropsWithoutRef,
-  type ElementRef,
   forwardRef,
+  useEffect,
+  useImperativeHandle,
+  useRef,
+  useState,
 } from 'react';
 
-import type { FormControlChildProps } from '@/Form';
-import { Input } from '@/lib/components/ui/input';
 import { cn } from '@/lib/utils';
+import { Button } from '../lib/components/ui/button';
+import { Calendar } from '../lib/components/ui/calendar';
+import {
+  Popover,
+  PopoverContent,
+  PopoverTrigger,
+} from '../lib/components/ui/popover';
+import {
+  type Granularity,
+  TimePicker12h,
+  TimePicker24h,
+} from '../lib/components/ui/time-picker';
+
+type DatetimeInputRef = {
+  value?: string;
+} & Omit<HTMLButtonElement, 'value'>;
+
+export interface DatetimeInputProps {
+  value?: string;
+  onChange?: (date: string | undefined) => void;
+  granularity?: Granularity;
+  hourCycle?: 12 | 24;
+}
 
 export const DatetimeInput = forwardRef<
-  ElementRef<typeof Input>,
-  FormControlChildProps & ComponentPropsWithoutRef<typeof Input>
->(({ value = '', className, ...inputProps }, ref) => {
-  if (typeof value !== 'string') {
-    throw new Error('value must be of type string.');
-  }
-  return (
-    <Input
-      type="datetime-local"
-      value={value?.substring(0, 19) ?? ''}
-      className={cn('dark:[color-scheme:dark]', className)}
-      ref={ref}
-      {...inputProps}
-    />
-  );
-});
+  Partial<DatetimeInputRef>,
+  DatetimeInputProps
+>(
+  (
+    { value, onChange, granularity = 'Second', hourCycle = 12, ...props },
+    ref,
+  ) => {
+    const [date, setDate] = useState<Date | undefined>(
+      value && !Number.isNaN(new Date(value)) ? new Date(value) : undefined,
```

Comment:
Check out what[date returns](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/Date#return_value). You need to include `valueOf()`